coderd/unhanger/detector.go

```diff
@@ -0,0 +1,332 @@
+			return xerrors.Errorf("acquire lock: %w", err)
+		}
+		d.log.Info(ctx, "running workspace build hang detector")
+
+		// Find all provisioner jobs that are currently running but have not
+		// received an update in the last 5 minutes.
+		jobs, err := db.GetHungProvisionerJobs(ctx, t.Add(-HungJobDuration))
+		if err != nil {
+			return xerrors.Errorf("get hung provisioner jobs: %w", err)
+		}
+
+		// Send a message into the build log for each hung job saying that it
+		// has been detected and will be terminated, then mark the job as
+		// failed.
+		for _, job := range jobs {
+			log := d.log.With(slog.F("job_id", job.ID))
+
+			jobStatus := db2sdk.ProvisionerJobStatus(job)
+			if jobStatus != codersdk.ProvisionerJobRunning {
+				log.Error(ctx, "hang detector query discovered non-running job, this is a bug", slog.F("status", jobStatus))
+				continue
+			}
+
+			log.Info(ctx, "detected hung (>5m) provisioner job, forcefully terminating")
+
+			// First, get the latest logs from the build so we can make sure
+			// our messages are in the latest stage.
+			logs, err := db.GetProvisionerLogsAfterID(ctx, database.GetProvisionerLogsAfterIDParams{
+				JobID:        job.ID,
+				CreatedAfter: 0,
+			})
+			if err != nil {
+				log.Warn(ctx, "get logs for hung job", slog.Error(err))
+				continue
+			}
+			logStage := ""
+			if len(logs) != 0 {
+				logStage = logs[len(logs)-1].Stage
+			}
+			if logStage == "" {
+				logStage = "Unknown"
+			}
+
+			// Insert the messages into the build log.
+			insertParams := database.InsertProvisionerJobLogsParams{
+				JobID: job.ID,
+			}
+			now := database.Now()
+			for i, msg := range HungJobLogMessages {
+				// Set the created at in a way that ensures each message has
+				// a unique timestamp so they will be sorted correctly.
+				insertParams.CreatedAt = append(insertParams.CreatedAt, now.Add(time.Millisecond*time.Duration(i)))
+				insertParams.Level = append(insertParams.Level, database.LogLevelError)
+				insertParams.Stage = append(insertParams.Stage, logStage)
+				insertParams.Source = append(insertParams.Source, database.LogSourceProvisionerDaemon)
+				insertParams.Output = append(insertParams.Output, msg)
+			}
+			newLogs, err := db.InsertProvisionerJobLogs(ctx, insertParams)
+			if err != nil {
+				log.Warn(ctx, "insert logs for hung job", slog.Error(err))
+				continue
+			}
+
+			// Publish the new log notification to pubsub. Use the lowest
+			// log ID inserted so the log stream will fetch everything after
+			// that point.
+			lowestID := newLogs[0].ID
+			data, err := json.Marshal(provisionersdk.ProvisionerJobLogsNotifyMessage{
+				CreatedAfter: lowestID - 1,
+			})
+			if err != nil {
+				log.Warn(ctx, "marshal log notification", slog.Error(err))
+				continue
+			}
+			err = d.pubsub.Publish(provisionersdk.ProvisionerJobLogsNotifyChannel(job.ID), data)
+			if err != nil {
+				log.Warn(ctx, "publish log notification", slog.Error(err))
+				continue
```

Comment:
Because I want to hold a lock I cannot use multiple transactions as the in memory DB only supports a single transaction at a time, and locks must be held within a transaction.

Reply:
Another point is that if publishing events fails it likely means the connection to the DB is down or the DB itself is having issues, so marking the job as complete has a high chance of failure anyways.

Once the DB is reachable again, the job will be picked up again and this process will be attempted again.

Reply:
I understand why a single transaction is preferable, but perhaps I didn't properly explain the problem I observed. Since the pubsub is not tied to the transaction, doing a publish inside a transaction that can still fail is a bit problematic. We will tell consumers there are logs after an ID that doesn't exist, for instance.

> I decided to do the log publishing before marking as complete because it seems weird to have logs flow in after the job is marked finished (and I feel that would break streaming etc.).

Since this is a transaction, the change is essentially atomic. Logs and "mark complete" will be done simultaneously. If it affects streaming we probably need to take a look at that.

Reply:
PS. I was suggesting sequential transactions per-job. Each would retry acquiring the lock and check that the job is still hung.

Reply:
I've changed to a sequential transaction per job, with a lock for each job.